From d72f436b7d7c697b262968c48c2d7643069ab17f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ond=C5=99ej=20Sur=C3=BD?= <ondrej@sury.org>
Date: Wed, 17 Apr 2019 15:22:27 +0200
Subject: [PATCH] Replace atomic operations in bin/named/client.c with
 isc_refcount reference counting

(cherry picked from commit ef49780d30d3ddc5735cfc32561b678a634fa72f)

Upstream-Status: Backport
https://gitlab.isc.org/isc-projects/bind9/commit/d72f436b7d7c697b262968c48c2d7643069ab17f#4bde2b1699e0b594852ceb7aa33df86998e664d0
Fixed up for 9.11 context

Signed-off-by: Armin Kuster <akuster@mvista.com>
---
 lib/ns/client.c                  | 18 +++++++-----------
 lib/ns/include/ns/interfacemgr.h |  5 +++--
 lib/ns/interfacemgr.c            |  7 +++++--
 3 files changed, 15 insertions(+), 15 deletions(-)

Index: bind-9.11.5-P4/bin/named/client.c
===================================================================
--- bind-9.11.5-P4.orig/bin/named/client.c
+++ bind-9.11.5-P4/bin/named/client.c
@@ -402,12 +402,10 @@ tcpconn_detach(ns_client_t *client) {
 static void
 mark_tcp_active(ns_client_t *client, bool active) {
 	if (active && !client->tcpactive) {
-		isc_atomic_xadd(&client->interface->ntcpactive, 1);
+		isc_refcount_increment0(&client->interface->ntcpactive, NULL);
 		client->tcpactive = active;
 	} else if (!active && client->tcpactive) {
-		uint32_t old =
-			isc_atomic_xadd(&client->interface->ntcpactive, -1);
-		INSIST(old > 0);
+		isc_refcount_decrement(&client->interface->ntcpactive, NULL);
 		client->tcpactive = active;
 	}
 }
@@ -554,7 +552,7 @@ exit_check(ns_client_t *client) {
 		if (client->mortal && TCP_CLIENT(client) &&
 		    client->newstate != NS_CLIENTSTATE_FREED &&
 		    !ns_g_clienttest &&
-		    isc_atomic_xadd(&client->interface->ntcpaccepting, 0) == 0)
+		    isc_refcount_current(&client->interface->ntcpaccepting) == 0)
 		{
 			/* Nobody else is accepting */
 			client->mortal = false;
@@ -3317,7 +3315,6 @@ client_newconn(isc_task_t *task, isc_eve
 	isc_result_t result;
 	ns_client_t *client = event->ev_arg;
 	isc_socket_newconnev_t *nevent = (isc_socket_newconnev_t *)event;
-	uint32_t old;
 
 	REQUIRE(event->ev_type == ISC_SOCKEVENT_NEWCONN);
 	REQUIRE(NS_CLIENT_VALID(client));
@@ -3337,8 +3334,7 @@ client_newconn(isc_task_t *task, isc_eve
 	INSIST(client->naccepts == 1);
 	client->naccepts--;
 
-	old = isc_atomic_xadd(&client->interface->ntcpaccepting, -1);
-	INSIST(old > 0);
+    isc_refcount_decrement(&client->interface->ntcpaccepting, NULL);
 
 	/*
 	 * We must take ownership of the new socket before the exit
@@ -3469,8 +3465,9 @@ client_accept(ns_client_t *client) {
 		 * quota is tcp-clients plus the number of listening
 		 * interfaces plus 1.)
 		 */
-		exit = (isc_atomic_xadd(&client->interface->ntcpactive, 0) >
-			(client->tcpactive ? 1 : 0));
+		 exit = (isc_refcount_current(&client->interface->ntcpactive) >
+            (client->tcpactive ? 1U : 0U));
+
 		if (exit) {
 			client->newstate = NS_CLIENTSTATE_INACTIVE;
 			(void)exit_check(client);
@@ -3528,7 +3525,7 @@ client_accept(ns_client_t *client) {
 	 * listening for connections itself to prevent the interface
 	 * going dead.
 	 */
-	isc_atomic_xadd(&client->interface->ntcpaccepting, 1);
+	isc_refcount_increment0(&client->interface->ntcpaccepting, NULL);
 }
 
 static void
Index: bind-9.11.5-P4/bin/named/include/named/interfacemgr.h
===================================================================
--- bind-9.11.5-P4.orig/bin/named/include/named/interfacemgr.h
+++ bind-9.11.5-P4/bin/named/include/named/interfacemgr.h
@@ -45,6 +45,7 @@
 #include <isc/magic.h>
 #include <isc/mem.h>
 #include <isc/socket.h>
+#include <isc/refcount.h>
 
 #include <dns/result.h>
 
@@ -75,11 +76,11 @@ struct ns_interface {
 						/*%< UDP dispatchers. */
 	isc_socket_t *		tcpsocket;	/*%< TCP socket. */
 	isc_dscp_t		dscp;		/*%< "listen-on" DSCP value */
-	int32_t			ntcpaccepting;	/*%< Number of clients
+	isc_refcount_t  ntcpaccepting;	/*%< Number of clients
 						     ready to accept new
 						     TCP connections on this
 						     interface */
-	int32_t			ntcpactive;	/*%< Number of clients
+	isc_refcount_t  ntcpactive;	/*%< Number of clients
 						     servicing TCP queries
 						     (whether accepting or
 						     connected) */
Index: bind-9.11.5-P4/bin/named/interfacemgr.c
===================================================================
--- bind-9.11.5-P4.orig/bin/named/interfacemgr.c
+++ bind-9.11.5-P4/bin/named/interfacemgr.c
@@ -386,8 +386,8 @@ ns_interface_create(ns_interfacemgr_t *m
 	 * connections will be handled in parallel even though there is
 	 * only one client initially.
 	 */
-	ifp->ntcpaccepting = 0;
-	ifp->ntcpactive = 0;
+    isc_refcount_init(&ifp->ntcpaccepting, 0);
+    isc_refcount_init(&ifp->ntcpactive, 0);
 
 	ifp->nudpdispatch = 0;
 
@@ -616,7 +616,8 @@ ns_interface_destroy(ns_interface_t *ifp
 
 	DESTROYLOCK(&ifp->lock);
 
-	ns_interfacemgr_detach(&ifp->mgr);
+	isc_refcount_destroy(&ifp->ntcpactive);
+    isc_refcount_destroy(&ifp->ntcpaccepting);
 
 	ifp->magic = 0;
 	isc_mem_put(mctx, ifp, sizeof(*ifp));
